src/buffer.cc

```diff
@@ -728,6 +728,50 @@ String Buffer::debug_description() const
                   content_size, additional_size);
 }
 
+InplaceString<23> to_string(Buffer::HistoryId id)
+{
+    if (id == Buffer::HistoryId::Invalid) {
+        InplaceString<23> res;
+        res.m_data[0] = '-';
+        res.m_length = 1;
+        return res;
+    } else {
+        return to_string(static_cast<size_t>(id));
+    }
+}
+
+String Buffer::format_modification(const Modification& modification, Quoting quoting) {
+    auto quote = quoter(quoting);
+    return quote(format("{}|{}.{}|{}",
+                        modification.type == Buffer::Modification::Type::Insert ? '+' : '-',
+                        modification.coord.line, modification.coord.column,
+                        modification.content->strview()));
+}
+
+String Buffer::history_as_string(Quoting quoting) const
```

Comment:
I dont think those should be in Buffer itself, I would prefer for buffer to expose a getter for history and uncommited modifications, and have free functions in buffer_utils.hh that format those to strings.

(Its okay to make Modification a public type in Buffer in order to achieve that).